Modified 360to180 method so that the overlap is now composed of both …

feinfinger opened the pull request:
…parts and weigthed with distance to the overlap center. This reduces artefacts in the center of the reconstruction.

dgursoy commented:
Is this failing because the function works with improved accuracy? If this is the case, could you also modify the function test so that it passes?

dmpelt commented:
Cool! I was planning to implement something similar: it makes sense to make as much use of the data as possible.

I have a few minor comments:

* Currently, the code will modify the incoming `data` variable: we probably want to avoid this (the user might use the data later [e.g. trying different overlap values or something]). One solution would be to apply the weighting when copying the data to the `out` array instead of inplace on the `data` array.
* The weighting now runs from 0 to 1, inclusive. Therefore, the values on the edge of the overlapping region are not used (they get a weight of zero). It is probably better to use all values (especially in low overlap cases), although the difference is probably very small.

So the code can become something like (untested):

```python
if rotation == 'left':
    weights = (np.arange(overlap)+0.5)/overlap
    out[:, :, -dz+overlap:] = data[:n, :, overlap:]
    out[:, :, :dz-overlap] = data[n:2*n, :, overlap:][:, :, ::-1]
    out[:, :, dz-overlap:dz] = weights*data[:n, :, :overlap] + (weights*data[n:2*n, :, :overlap])[:, :, ::-1]
elif rotation == 'right':
    weights = (np.arange(overlap)[::-1]+0.5)/overlap
    out[:, :, :dz-overlap] = data[:n, :, :-overlap]
    out[:, :, -dz+overlap:] = data[n:2*n, :, :-overlap][:, :, ::-1]
    out[:, :, dz-overlap:dz] = weights*data[:n, :, -overlap:] + (weights*data[n:2*n, :, -overlap:])[:, :, ::-1]
```

feinfinger commented:
@Doga On behalf of the test: yes, the test fails, because the overlap
too is compared to the input array, which would be ok, if the data had
the symmetry a 360deg scan should have. The test uses a random array, so
the symmetry is not given. I will work on the test as soon as I can
(probably next week).

@Daniel :-) thanks. I have to admit I don't quite get the first point,
since the data array lives only in the function scope, right? So the
input data from wherever it is called is still untouched. As to the
second point, it would be easy to implement kwargs for the weighting
parameters, although the clarity of the function might suffer from that.
Another possibility would be to make weighting altogether optional.


On 2016-12-08 01:12, Daniel M. Pelt wrote:
>
> Cool! I was planning to implement something similar: it makes sense to
> make as much use of the data as possible.
>
> I have a few minor comments:
>
>   * Currently, the code will modify the incoming |data| variable: we
>     probably want to avoid this (the user might use the data later
>     [e.g. trying different overlap values or something]). One solution
>     would be to apply the weighting when copying the data to the |out|
>     array instead of inplace on the |data| array.
>   * The weighting now runs from 0 to 1, inclusive. Therefore, the
>     values on the edge of the overlapping region are not used (they
>     get a weight of zero). It is probably better to use all values
>     (especially in low overlap cases), although the difference is
>     probably very small.
>
> So the code can become something like (untested):
>
> if rotation == 'left':
>     weights = (np.arange(overlap)+0.5)/overlap
>     out[:, :, -dz+overlap:] = data[:n, :, overlap:]
>     out[:, :, :dz-overlap] = data[n:2*n, :, overlap:][:, :, ::-1]
>     out[:, :, dz-overlap:dz] = weights*data[:n, :, :overlap] + (weights*data[n:2*n, :, :overlap])[:, :, ::-1]
> elif rotation == 'right':
>     weights = (np.arange(overlap)[::-1]+0.5)/overlap
>     out[:, :, :dz-overlap] = data[:n, :, :-overlap]
>     out[:, :, -dz+overlap:] = data[n:2*n, :, :-overlap][:, :, ::-1]
>     out[:, :, dz-overlap:dz] = weights*data[:n, :, -overlap:] + (weights*data[n:2*n, :, -overlap:])[:, :, ::-1]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tomopy/tomopy/pull/240#issuecomment-265614587>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcnMS41Dxaz37MmaXdr957wzS6NWm5rks5rF0tfgaJpZM4LCZub>.
>

-- 
Dr. Fabian Wilde
Helmholtz-Zentrum Geesthacht
Institute for Materials Research
c/o HZG at DESY - Notkestr. 85 - 22607 Hamburg - Germany
Phone: +49 (0) 40 8998 5334
Fax:  +49 (0) 40 8994 5334
Email: fabian.wilde@hzg.de

Helmholtz-Zentrum Geesthacht 
Zentrum für Material- und Küstenforschung GmbH 
Max-Planck-Straße 1 I 21502 Geesthacht I Deutschland/Germany 

Geschäftsführer/Board of Management: Prof. Dr. Wolfgang Kaysser, Dipl.-Ing. Michael Ganß 
Vorsitzender des Aufsichtsrates/Chairman of the Supervisory Board: Ministerialdirigent Dr. Herbert Zeisel 
Amtsgericht Lübeck HRB 285 GE (Register Court) 
Internet: http://www.hzg.de

dmpelt commented:
@feinfinger About the first point: the `data` reference is indeed local to the function scope, but the underlying object (and the associated array) is not local to the scope, and will be changed. Python's way of argument passing is a bit different from C-like languages, but generally you can think of it as pass-by-reference (although this is not technically correct, see [here](https://jeffknupp.com/blog/2012/11/13/is-python-callbyvalue-or-callbyreference-neither/)). Anyway, the following example shows the behavior in the case of this PR:

```python
In [1]: tst = [1,2,3]

In [2]: def change(data):
   ...:     data[0] = 10
   ...:     

In [3]: change(tst)

In [4]: tst
Out[4]: [10, 2, 3]
```

About the second point: I like the idea of having different weightings to choose from. For example, only using the data that is furthest from the edge of the detector can be a good idea when there are significant artifacts at the detector edge.

feinfinger commented:
@Daniel: yes of course - you are right! I tested more or less the same thing with a =a*2 in the function before I wrote my email, which leaves the initial data intact (and is in that different to a*=2). I’ll try to figure out a memory efficient way around this issue.

> On 09.12.2016, at 20:34, Daniel M. Pelt <notifications@github.com> wrote:
> 
> @feinfinger <https://github.com/feinfinger> About the first point: the data reference is indeed local to the function scope, but the underlying object (and the associated array) is not local to the scope, and will be changed. Python's way of argument passing is a bit different from C-like languages, but generally you can think of it as pass-by-reference (although this is not technically correct, see here <https://jeffknupp.com/blog/2012/11/13/is-python-callbyvalue-or-callbyreference-neither/>). Anyway, the following example shows the behavior in the case of this PR:
> 
> In [1]: tst = [1,2,3]
> 
> In [2]: def change(data):
>    ...:     data[0] = 10
>    ...:     
> 
> In [3]: change(tst)
> 
> In [4]: tst
> Out[4]: [10, 2, 3]
> About the second point: I like the idea of having different weightings to choose from. For example, only using the data that is furthest from the edge of the detector can be a good idea when there are significant artifacts at the detector edge.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/tomopy/tomopy/pull/240#issuecomment-266101902>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAcnMdjUrsFFVE8ziVp0Qzl_hVMFErbnks5rGa0xgaJpZM4LCZub>.
> 

Helmholtz-Zentrum Geesthacht 
Zentrum für Material- und Küstenforschung GmbH 
Max-Planck-Straße 1 I 21502 Geesthacht I Deutschland/Germany 

Geschäftsführer/Board of Management: Prof. Dr. Wolfgang Kaysser, Dipl.-Ing. Michael Ganß 
Vorsitzender des Aufsichtsrates/Chairman of the Supervisory Board: Ministerialdirigent Dr. Herbert Zeisel 
Amtsgericht Lübeck HRB 285 GE (Register Court) 
Internet: http://www.hzg.de

feinfinger commented:
Sorry for not having worked on this modification for so long. I now followed Daniels implementation and modified the test routine. For now I have implemented just on estandard weighting to keep things simple.
